TW 332 API security admin panel & API governance admin panel (#4243)
postmanlabs/postman-docs: src/pages/docs/api-governance/configurable-rules/spectral.md

---
title: "Using Spectral in Postman"
updated: 2022-09-15
---

Spectral is a linting engine that helps you define custom rules and enforce them on JSON and YAML files. Postman supports Spectral v6 rules for the configured [API Governance](/docs/api-governance/configurable-rules/configuring-api-governance-rules/#adding-custom-rules) and [API Security](/docs/api-governance/configurable-rules/configuring-api-security-rules/#adding-custom-rules) rules for your team.

## Contents

* [How Spectral works](#how-spectral-works)
    * [Rulesets in Spectral documents](#rulesets-in-spectral-documents)
* [Spectral support in Postman](#spectral-support-in-postman)
* [Spectral rule properties](#spectral-rule-properties)
* [JSON Path and JSON Path Plus](#json-path-and-json-path-plus)

>     * [Building and testing JSON Path Plus expressions](building-and-testing-json-path-plus-expressions)

Reviewer: ```suggestion
    * [Building and testing JSON Path Plus expressions](#building-and-testing-json-path-plus-expressions)
```

    * [JSON Path Plus examples](#json-path-plus-examples)
* [Example: Checking for the presence of a property](#example-checking-for-the-presence-of-a-property)

## How Spectral works

Spectral checks that the APIs defined in OpenAPI documents conform to API design guidelines using a specific set of rules. For example, with Spectral you can check that all properties of all data models are in camel case, or that all operations have a summary.

> A Spectral rule targets a given location using a [JSON Path Plus expression](#json-path-and-json-path-plus), then tests the values it finds with a `function`. It returns an error if the values don't conform to the rule.

Reviewer: Does function need to be formatted as an inline code block?

Author: Double-checked and yep, it's supposed to be. Which I missed on line 37!

This example shows how to check that the name of an API doesn't contain the word "API," regardless of case (for instance, "api" or "Api").

```
rules:
  api-name-doesnt-contain-api:
    given: $.info.title
    then:
      function: pattern
      functionOptions:
          notMatch: /\b(api)\b/i
```

> First, the rule targets the `title` property of the `info` object located at the root (`$`) of the document with the [JSON Path Plus expression](#json-path-and-json-path-plus) `$.info.title`. In `then`, the function named `pattern` checks whether the value of the title property doesn't match (`functionOptions.notMatch`) the regular expression looking for the word "api" in any case (`/\b(api)\b/i`).

Reviewer: ```suggestion
First, the rule targets the `title` property of the `info` object located at the root (`$`) of the document with the [JSON Path Plus expression](#json-path-and-json-path-plus) `$.info.title`. In `then`, the function named `pattern` checks whether the value of the `title` property doesn't match (`functionOptions.notMatch`) the regular expression looking for the word "api" in any case (`/\b(api)\b/i`).
```

> First, the rule targets the `title` property of the `info` object located at the root (`$`) of the document with the [JSON Path Plus expression](#json-path-and-json-path-plus) `$.info.title`. In `then`, the function named `pattern` checks whether the value of the `title` property doesn't match (`functionOptions.notMatch`) the regular expression looking for the word "api" in any case (`/\b(api)\b/i`).

Author: ```suggestion
First, the rule targets the `title` property of the `info` object located at the root (`$`) of the document with the [JSON Path Plus expression](#json-path-and-json-path-plus) `$.info.title`. In `then`, the `function` named `pattern` checks whether the value of the `title` property doesn't match (`functionOptions.notMatch`) the regular expression looking for the word "api" in any case (`/\b(api)\b/i`).
```

## Spectral support in Postman

Postman supports many of Spectral's features, though not all.

Whether it's created within Postman or imported from another source, a Spectral document needs to contain the properties shown in the example below:

```
rules:
  api-name-doesnt-contain-api:
    description: The API name must not contain the word API
    message: The info.title value "{{value}}" contains the forbidden word API
    severity: error
    formats:
      - oas2
      - oas3
    given:
      - $.info
    then:
      - field: title
        function: pattern
        functionOptions:
          notMatch: /\b(api)\b/i
```

To be valid in Postman, your Spectral document can't contain any properties beyond those listed here. For the full list of rules and their descriptions, see [Spectral rule properties](#spectral-rule-properties).

You will find each rule defined in `rules` in the **Custom Rules** section in the configurable [API Governance](/docs/api-governance/configurable-rules/configuring-api-governance-rules/) or [API Security](/docs/api-governance/configurable-rules/configuring-api-security-rules/) pages.

> > Don't forget to turn on your custom rules after you create or import them!

Reviewer: It's my understanding that we typically avoid exclamation points. Also, instead of "Don't forget", how about "Remember"? To me, this is easier to scan.

**Proposal:**
`Remember to turn on your custom rules after you create or import them.`

> > Don't forget to turn on your custom rules after you create or import them!

Author: ```suggestion
> Remember to turn on your custom rules after you create or import them.
```

## Spectral rule properties

|<div style="width:110px">Property</div> | Description
--- | ---
`description` | An optional description of the rule. If you provide one, it will be shown in the configurable rules page for either API Governance or API Security.
`message` | If the rule is triggered, the list of rule violations will contain the `message`, used in Postman as the name of the rule. This message aims to help users solve the problem. Keep it as short and meaningful as possible. It can contain optional placeholders: <br><ul><li>`{{error}}` - The error message returned by `function`.</li><li>`{{description}}` - The description of the rule.</li><li>`{{path}}` - The path of the error (the last element is the `property` below).</li><li>`{{property}}` - The name of the property causing the error. This is useful when `given` returns many different property names or when `then` is a list that uses multiple `fields`).</li><li>`{{value}}` - The value causing the error.</li></ul><br> If `message` isn't provided, the `description` is used instead. And if `description` isn't available, the rule's key (in `rules`) is used.

> `severity` | The severity of the problem detected by the rule. The possible values are `error`, `warn` (default), `info`, and `hint`. These values can be used as follows: <br><ul><li>`error` - An obvious error that must be fixed. Something is going against guidelines.</li><li>`warn` - A possible error. If it's an error, it must be fixed. Some deviation on specific rules may be tolerated, like a `POST` without a body.</li><li>`info`: Something that could possibly be improved. An optional pattern defined in the guidelines could be applied.</li><li>`hint` - Something to be discussed during an API design review.</li></ul>

Reviewer: ```suggestion
`severity` | The severity of the problem detected by the rule. The possible values are `error`, `warn` (default), `info`, and `hint`. These values can be used as follows: <br><ul><li>`error` - An obvious error that must be fixed. Something is going against guidelines.</li><li>`warn` - A possible error. If it's an error, it must be fixed. Some deviation on specific rules may be tolerated, like a `POST` without a body.</li><li>`info` - Something that could possibly be improved. An optional pattern defined in the guidelines could be applied.</li><li>`hint` - Something to be discussed during an API design review.</li></ul>
```

`formats` | The list of document formats to which the rule will be applied. The accepted values are: <br><ul><li>`oas2` - Targets OpenAPI (Swagger) documents</li><li>`oas3` - Targets OpenAPI 3.x documents (3.0 and 3.1)</li><li>`oas3_0` - Targets OpenAPI 3.0 documents</li><li>`oas3_1` - Targets OpenAPI 3.1 documents</li></ul><br> By default, a rule will target all versions 2 and 3.x (the default value is `[oas2,oas3]`).
`given` | **Required**. This can be a list with at least one element or a single element. Each value is a [JSON Path Plus expression](#json-path-and-json-path-plus) that may return zero, one, or more elements.<br>If `given` paths don't find any value, the `then` controls won't execute.

> `then` | **Required**. This can be a list with at least one element or a single element. If the given JSON Path Plus expressions return values, the functions will be applied to all of them.

Reviewer: Should "JSON Path Plus expressions" also include an anchor link to the section? I'm asking because it looks like we're doing that in other rows.

> `then` | **Required**. This can be a list with at least one element or a single element. If the given JSON Path Plus expressions return values, the functions will be applied to all of them.

Author: ```suggestion
`then` | **Required**. This can be a list with at least one element or a single element. If the given [JSON Path Plus expressions](#json-path-and-json-path-plus) return values, the functions will be applied to all of them.
```

Initially, JSON Path was created to be [XPath for JSON](https://goessner.net/articles/JsonPath/) and aimed to represent the path to some element in an XML document. [JSON Path Plus](https://jsonpath-plus.github.io/JSONPath/docs/ts/) is an extension of JSON Path. It adds more operators and makes some behaviors of the original specification explicit.

### Building and testing JSON Path Plus expressions

You can use the official JSON Path Plus [documentation](https://jsonpath-plus.github.io/JSONPath/docs/ts/) (especially [Syntax Through Examples](https://jsonpath-plus.github.io/JSONPath/docs/ts/#syntax-through-examples)) and the [JSON Path Plus demo](https://jsonpath-plus.github.io/JSONPath/demo/) to build and test your rules' given paths.

### JSON Path Plus examples

These examples show the typical JSON Path Plus features you'll need to create rules in Postman:

* `$.info.title` - The `title` property inside the `info` object, which is located at the document's root (`$`).
* `$.paths.*.*.responses` - All responses of all operations of all `paths`. `*` gets all elements inside an object or an array.

> * `$.paths.*[post,patch,put]` - All POST, PATCH, and PUT operations across all `paths`. `[ ]` can be used to filter elements.

Reviewer: ```suggestion
* `$.paths.*[post,patch,put]` - All `POST`, `PATCH`, and `PUT` operations across all `paths`. `[ ]` can be used to filter elements.
```

---

> title: "Using Spectral in Postman"

Author: @savageCoder55 @arno-di-loreto @chinmayjoshi This is the new Spectral page, based on Arnaud's draft. Please take a look!

@MatthewPinsker-postman and/or @jkonrath-postman if you could also take a look from a TW perspective I'd appreciate it 🙏

Reviewer: @HeatherStenson Looks good to me! I added a few proposed changes.

Author: Thank you!!!

updated: 2022-09-15
---

Spectral is a linting engine that helps you define custom rules and enforce them on JSON and YAML files. Postman supports Spectral v6 rules for the configured [API Governance](/docs/api-governance/configurable-rules/configuring-api-governance-rules/) and [API Security](/docs/api-governance/configurable-rules/configuring-api-security-rules/) rules for your team.

> [Team Admins](/docs/collaborating-in-postman/roles-and-permissions/) can define rules according to Spectral guidelines, then import them into Postman as custom rules.

Reviewer 2: Let's just link back to the create rule pages in API Governance and API Security sections.
This would be cleaner in my opinion.

Just a thought.

## Contents

* [How Spectral works](#how-spectral-works)
    * [Rulesets in Spectral documents](#rulesets-in-spectral-documents)
* [Spectral support in Postman](#spectral-support-in-postman)
* [Spectral rule properties](#spectral-rule-properties)
* [JSON Path and JSON Path Plus](#json-path-and-json-path-plus)
    * [Building and testing JSON Path Plus expressions](building-and-testing-json-path-plus-expressions)
    * [JSON Path Plus examples](#json-path-plus-examples)
* [Example: Checking for the presence of a property](#example-checking-for-the-presence-of-a-property)

## How Spectral works

> Spectral is a linter that checks whether a JSON or YAML document conforms to a set of defined rules. It's used to check that the APIs defined in OpenAPI documents conform to API design guidelines specified in the context of API governance policies. For example, with Spectral you can check that all properties of all data models are in camel case, or that all operations have a summary.

Reviewer 2: This first sentence I believe is not needed - we cover this in the intro to this page as well.

Reviewer 2: instead of `API design guidelines specified in the context of API governance policies` I'd suggest replacing with `API design guidelines against a set of rules`

This example shows how to check that the name of an API doesn't contain the word "API," regardless of case (for instance, "api" or "Api").

```
rules:
  api-name-doesnt-contain-api:
    given: $.info.title
    then:
      function: pattern
      functionOptions:
          notMatch: /\b(api)\b/i
```

First, the rule targets the `title` property of the `info` object located at the root (`$`) of the document with the [JSON Path Plus expression](#json-path-and-json-path-plus) `$.info.title`. In `then`, the function named `pattern` checks whether the value of the title property doesn't match (`functionOptions.notMatch`) the regular expression looking for the word "api" in any case (`/\b(api)\b/i`).

### Rulesets in Spectral documents

A Spectral document (often called a ruleset) contains a `rules` property that can have one or more rules.

```
rules:
api-name-doesnt-contain-api:
# ...
api-name-length:
# ...
```

Spectral identifies each rule with a key, which can be any JSON-compatible string.

## Spectral support in Postman

Postman supports many of Spectral's features, though not all.

> Whether it's created within Postman or imported from another source, a Spectral document needs to contain the following properties:

Reviewer 2: at the end, let's say that `a Spectral document needs to contain the properties as shown in the example below`

```
rules:
  api-name-doesnt-contain-api:
    description: The API name must not contain the word API
    message: The info.title value "{{value}}" contains the forbidden word API
    severity: error
    formats:
      - oas2
      - oas3
    given:
      - $.info
    then:
      - field: title
        function: pattern
        functionOptions:
          notMatch: /\b(api)\b/i
```

To be valid in Postman, your Spectral document can't contain any properties beyond those listed here. For the full list of rules and their descriptions, see [Spectral rule properties](#spectral-rule-properties).

You will find each rule defined in `rules` in the **Custom Rules** section in the configurable [API Governance](/docs/api-governance/configurable-rules/configuring-api-governance-rules/) or [API Security](/docs/api-governance/configurable-rules/configuring-api-security-rules/) pages.

> Don't forget to turn on your custom rules after you create or import them!

## Spectral rule properties

Property | Description
--- | ---

> `description` | An optional description of the rule. It will be shown in the API governance settings you provide one.

Reviewer 2: instead of `API Governance settings` let's say the `Configurable rules page for either API Governance or API Security`
